rust/src/storage/s3/dynamodb_lock.rs

```diff
@@ -0,0 +1,510 @@
+    /// The expression that checks whether the lock record does not exists.
+    pub const ACQUIRE_LOCK_THAT_DOESNT_EXIST: &str = "attribute_not_exists(#pk)";
+
+    /// The expression that checks whether the lock record exists and it is marked as released.
+    pub const PK_EXISTS_AND_IS_RELEASED: &str = "attribute_exists(#pk) AND #ir = :ir";
+
+    /// The expression that checks whether the lock record exists
+    /// and its record version number matches with the given one.
+    pub const PK_EXISTS_AND_RVN_MATCHES: &str = "attribute_exists(#pk) AND #rvn = :rvn";
+
+    /// The expression that checks whether the lock record exists,
+    /// its record version number matches with the given one
+    /// and its owner name matches with the given one.
+    pub const PK_EXISTS_AND_OWNER_RVN_MATCHES: &str =
+        "attribute_exists(#pk) AND #rvn = :rvn AND #on = :on";
+
+    /// The expression that updates is_released and data fields of DynamoDB record.
+    pub const UPDATE_IS_RELEASED_AND_DATA: &str = "SET #ir = :ir, #d = :d";
+
+    /// The expression that updates is_released field of DynamoDB record.
+    pub const UPDATE_IS_RELEASED: &str = "SET #ir = :ir";
+}
+
+mod vars {
+    pub const PK_PATH: &str = "#pk";
+    pub const RVN_PATH: &str = "#rvn";
+    pub const RVN_VALUE: &str = ":rvn";
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+    pub const OWNER_NAME_PATH: &str = "#on";
+    pub const OWNER_NAME_VALUE: &str = ":on";
+    pub const DATA_PATH: &str = "#d";
+    pub const DATA_VALUE: &str = ":d";
+}
+
+/// Provides a simple library for using DynamoDB's consistent read/write feature
+/// to use it for managing distributed locks.
+pub struct DynamoDbLockClient {
+    client: DynamoDbClient,
+    opts: Options,
+}
+
+impl DynamoDbLockClient {
+    /// Creates new DynamoDB lock client
+    pub fn new(client: DynamoDbClient, opts: Options) -> Self {
+        Self { client, opts }
+    }
+
+    /// Attempts to acquire a lock until it either acquires the lock or a specified
+    /// `additional_time_to_wait_for_lock` is reached. This function will poll DynamoDB based
+    /// on the `refresh_period`. If it does not see the lock in DynamoDB, it will immediately
+    /// return the lock to the caller. If it does see the lock, it will note the lease
+    /// expiration on the lock. If the lock is deemed stale, then this will acquire and return it.
+    /// Otherwise, if it waits for as long as `additional_time_to_wait_for_lock` without acquiring
+    /// the lock, then it will a [`DynamoError::TimedOut].
+    ///
+    /// Note that this method will wait for at least as long as the `lease_duration` in order
+    /// to acquire a lock that already exists. If the lock is not acquired in that time,
+    /// it will wait an additional amount of time specified in `additional_time_to_wait_for_lock`
+    /// before giving up.
+    pub async fn acquire_lock(&self) -> Result<LockItem, DynamoError> {
```

Comment:
@houqp Okay so there's no `Option` yet as in original's but here's some thoughts. There's 2 expected ways of exiting a acquire loop, [someone else acquired the lock](https://github.com/awslabs/amazon-dynamodb-lock-client/blob/master/src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java#L532)  or [provisioned throughput exceeded](https://github.com/awslabs/amazon-dynamodb-lock-client/blob/master/src/main/java/com/amazonaws/services/dynamodbv2/AmazonDynamoDBLockClient.java#L538).
However they use single exception for those cases and a new method `public Optional<LockItem> tryAcquireLock` to try-catch on that and return None if `LockNotGrantedException` is caught.  

We might however still use `TimedOut` and new one for "provisioned throughput exceeded" to pattern match for `Option<LockItem>`.

wdyt?

Reply:
yep, i think your design is much better than the reference implementation :) I am not a fan of exception anyway, not to mention a single catch all exception :P